rust-lldb: fix crash when printing empty string

euclio opened the pull request:
Fixes #52185.

~Re-enables the pretty-std debuginfo test and tweaks the test as necessary to get it to pass again. This reveals that lldb's formatting of enums is broken (#58492). I also removed the emoji from the test because I couldn't get the docker image's gdb to print the emoji, just octal escapes (https://github.com/rust-lang/rust/pull/53154/files#r208263904).~

rust-highfive commented:
r? @alexcrichton

(rust_highfive has picked a reviewer for you, use r? to override)

rust-highfive commented:
r? @nikomatsakis

(rust_highfive has picked a reviewer for you, use r? to override)

rust-highfive commented:
The job `x86_64-gnu-llvm-6.0` of your PR [failed on Travis](https://travis-ci.com/rust-lang/rust/jobs/179700324) ([raw log](https://api.travis-ci.com/v3/job/179700324/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
travis_time:end:29dee5c6:start=1550798865224776861,finish=1550798866250241594,duration=1025464733
$ git checkout -qf FETCH_HEAD
travis_fold:end:git.checkout

Encrypted environment variables have been removed for security reasons.
See https://docs.travis-ci.com/user/pull-requests/#pull-requests-and-security-restrictions
$ export SCCACHE_BUCKET=rust-lang-ci-sccache2
$ export SCCACHE_REGION=us-west-1
Setting environment variables from .travis.yml
$ export IMAGE=x86_64-gnu-llvm-6.0
---
travis_time:start:test_debuginfo
Check compiletest suite=debuginfo mode=debuginfo-both (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
[01:13:16] 
[01:13:16] running 119 tests
[01:13:45] .iiiii...i.....i..i...i..i.i..i.ii...i.....i..i....i..........iiii..........i...iF...i.......ii.i.i. 100/119
[01:13:50] i......iii.i.....ii
[01:13:50] 
[01:13:50] ---- [debuginfo-both] debuginfo/pretty-std.rs stdout ----
[01:13:50] ---- [debuginfo-both] debuginfo/pretty-std.rs stdout ----
[01:13:50] NOTE: compiletest thinks it is using GDB without native rust support
[01:13:50] NOTE: compiletest thinks it is using GDB version 7011001
[01:13:50] 
[01:13:50] error: line not found in debugger output: $5 = core::option::Option<i16>::Some(8)
[01:13:50] status: exit code: 0
[01:13:50] command: "/usr/bin/gdb" "-quiet" "-batch" "-nx" "-command=/checkout/obj/build/x86_64-unknown-linux-gnu/test/debuginfo/pretty-std/pretty-std.debugger.script"
[01:13:50] ------------------------------------------
[01:13:50] GNU gdb (Ubuntu 7.11.1-0ubuntu1~16.5) 7.11.1
[01:13:50] Copyright (C) 2016 Free Software Foundation, Inc.
[01:13:50] Copyright (C) 2016 Free Software Foundation, Inc.
[01:13:50] License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
[01:13:50] This is free software: you are free to change and redistribute it.
[01:13:50] There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
[01:13:50] and "show warranty" for details.
[01:13:50] This GDB was configured as "x86_64-linux-gnu".
[01:13:50] Type "show configuration" for configuration details.
[01:13:50] For bug reporting instructions, please see:
[01:13:50] <http://www.gnu.org/software/gdb/bugs/>.
[01:13:50] Find the GDB manual and other documentation resources online at:
[01:13:50] <http://www.gnu.org/software/gdb/documentation/>.
[01:13:50] For help, type "help".
[01:13:50] Type "apropos word" to search for commands related to "word".
[01:13:50] Breakpoint 1 at 0x31a3: file /checkout/src/test/debuginfo/pretty-std.rs, line 100.
[01:13:50] [Thread debugging using libthread_db enabled]
[01:13:50] Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
[01:13:50] 
[01:13:50] Breakpoint 1, pretty_std::main::hd868472729f88994 () at /checkout/src/test/debuginfo/pretty-std.rs:100
[01:13:50] 100     zzz(); // #break
[01:13:50] $1 = &[i32](len: 4) = {0, 1, 2, 3}
[01:13:50] $2 = Vec<u64>(len: 4, cap: 4) = {4, 5, 6, 7}
[01:13:50] $3 = "IAMA string slice!"
[01:13:50] $4 = "IAMA string!"
[01:13:50] $5 = Some = {8}
[01:13:50] $6 = None
[01:13:50] $7 = "IAMA OS string \360\237\230\203"
[01:13:50] $8 = {RUST$ENCODED$ENUM$0$None = Some = {"IAMA optional string!"}}
[01:13:50] $9 = {RUST$ENCODED$ENUM$0$None = Some = {"IAMA "...}}
[01:13:50] $10 = ""
[01:13:50] A debugging session is active.
[01:13:50] 
[01:13:50]  Inferior 1 [process 761] will be killed.
[01:13:50] 
[01:13:50] Quit anyway? (y or n) [answered Y; input not from terminal]
[01:13:50] ------------------------------------------
[01:13:50] stderr:
[01:13:50] ------------------------------------------
[01:13:50] ------------------------------------------
[01:13:50] Python Exception <class 'gdb.error'> Cannot convert value to long.: 
[01:13:50] Python Exception <class 'gdb.error'> Cannot convert value to long.: 
[01:13:50] Python Exception <class 'gdb.error'> Cannot convert value to long.: 
[01:13:50] Python Exception <class 'gdb.error'> Cannot convert value to long.: 
[01:13:50] ------------------------------------------
[01:13:50] 
[01:13:50] thread 'main' panicked at 'explicit panic', src/tools/compiletest/src/runtest.rs:3295:9
[01:13:50] note: Run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
---
[01:13:50] 
[01:13:50] thread 'main' panicked at 'Some tests failed', src/tools/compiletest/src/main.rs:496:22
[01:13:50] 
[01:13:50] 
[01:13:50] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--src-base" "/checkout/src/test/debuginfo" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/debuginfo" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--mode" "debuginfo-both" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/usr/lib/llvm-6.0/bin/FileCheck" "--host-rustcflags" "-Crpath -O -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--target-rustcflags" "-Crpath -O -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/usr/bin/python2.7" "--lldb-python" "/usr/bin/python2.7" "--gdb" "/usr/bin/gdb" "--quiet" "--llvm-version" "6.0.0\n" "--system-llvm" "--cc" "" "--cxx" "" "--cflags" "" "--llvm-components" "" "--llvm-cxxflags" "" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--color" "always"
[01:13:50] 
[01:13:50] 
[01:13:50] failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test
[01:13:50] Build completed unsuccessfully in 0:12:48
[01:13:50] Build completed unsuccessfully in 0:12:48
[01:13:50] make: *** [check] Error 1
[01:13:50] Makefile:48: recipe for target 'check' failed
The command "stamp sh -x -c "$RUN_SCRIPT"" exited with 2.
travis_time:start:13d41035
$ date && (curl -fs --head https://google.com | grep ^Date: | sed 's/Date: //g' || true)
Fri Feb 22 02:41:47 UTC 2019
---
travis_time:end:1f653120:start=1550803309094310077,finish=1550803309153947276,duration=59637199
travis_fold:end:after_failure.4
travis_fold:start:after_failure.5
travis_time:start:36499595
$ cat ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers || true
cat: ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers: No such file or directory
travis_fold:end:after_failure.5
travis_fold:start:after_failure.6
travis_time:start:0015bd3a
$ dmesg | grep -i kill
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

alexcrichton commented:
@bors: r+

cc @michaelwoerister (in case you'd like to double check)

bors commented:
:pushpin: Commit 7c74bace2e5db7da93bde03cd8761728719ee5f6 has been approved by `alexcrichton`

<!-- @bors r=alexcrichton 7c74bace2e5db7da93bde03cd8761728719ee5f6 -->

Centril commented:
*Probably* failed in https://github.com/rust-lang/rust/pull/58688#issuecomment-466732416, @bors r-

euclio commented:
Ok, since the pretty-std test looks like it's pretty flaky, I reverted the changes I made to it and extracted the empty string tests into their own file.

Centril commented:
r? @alexcrichton

alexcrichton commented:
@bors: r+

bors commented:
:pushpin: Commit b7d4f96d195323cf9344a2af6720dfae3d9f299d has been approved by `alexcrichton`

<!-- @bors r=alexcrichton b7d4f96d195323cf9344a2af6720dfae3d9f299d -->

bors commented:
:hourglass: Testing commit b7d4f96d195323cf9344a2af6720dfae3d9f299d with merge dfb9fd5eef9812fd1539b124885400eb0f8a87b8...

bors commented:
:broken_heart: Test failed - [checks-travis](https://travis-ci.com/rust-lang/rust/builds/102313327)

rust-highfive commented:
Your PR [failed on Travis](https://travis-ci.com/rust-lang/rust/jobs/180579404) ([raw log](https://api.travis-ci.com/v3/job/180579404/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
Attempting to download s3://rust-lang-ci-sccache2/docker/7e1a0f6dbf7ee5eb0f689d3f3f0b0c18b1a9c99aaacc2a649ce33a6db602dde1d880b783741255f4fb7da1afa8a7b1e7897822d59f692a8c90e11f791375e7c7
[00:00:10] Attempting with retry: curl -y 30 -Y 10 --connect-timeout 30 -f -L -C - -o /tmp/rustci_docker_cache https://s3-us-west-1.amazonaws.com/rust-lang-ci-sccache2/docker/7e1a0f6dbf7ee5eb0f689d3f3f0b0c18b1a9c99aaacc2a649ce33a6db602dde1d880b783741255f4fb7da1afa8a7b1e7897822d59f692a8c90e11f791375e7c7
[00:00:10]   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
[00:00:10]                                  Dload  Upload   Total   Spent    Left  Speed
No output has been received in the last 30m0s, this potentially indicates a stalled build or something wrong with the build itself.
Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#Build-times-out-because-no-output-was-received
The build has been terminated
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

alexcrichton commented:
@bors: retry

Centril commented:
Failed in https://github.com/rust-lang/rust/pull/58766#issuecomment-467760212, @bors r-

euclio commented:
Ignored windows.

alexcrichton commented:
@bors: r+

bors commented:
:pushpin: Commit 932fe17c41845c610e193c84b6947581190d782e has been approved by `alexcrichton`

<!-- @bors r=alexcrichton 932fe17c41845c610e193c84b6947581190d782e -->

bors commented:
:hourglass: Testing commit 932fe17c41845c610e193c84b6947581190d782e with merge 2b7bcefa602313686484007e88a31a25647bfe13...
<!-- homu: {"type":"BuildStarted","head_sha":"932fe17c41845c610e193c84b6947581190d782e","merge_sha":"2b7bcefa602313686484007e88a31a25647bfe13"} -->

bors commented:
:broken_heart: Test failed - [checks-travis](https://travis-ci.com/rust-lang/rust/builds/102953153)
<!-- homu: {"type":"BuildFailed","builder_url":"https://travis-ci.com/rust-lang/rust/builds/102953153","builder_name":"checks-travis"} -->

rust-highfive commented:
The job `arm-android` of your PR [failed on Travis](https://travis-ci.com/rust-lang/rust/jobs/181899068) ([raw log](https://api.travis-ci.com/v3/job/181899068/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
[01:42:09] 
[01:42:09] failures:
[01:42:09] thread 'main' panicked at 'Some tests failed', src/tools/compiletest/src/main.rs:496:22
[01:42:09] 
[01:42:09] ---- [debuginfo-both] debuginfo/empty-string.rs stdout ----
[01:42:09] NOTE: compiletest thinks it is using GDB without native rust support
[01:42:09] 
[01:42:09] error: line not found in debugger output: $1 = ""
[01:42:09] status: exit code: 0
[01:42:09] command: "arm-linux-androideabi-gdb" "-quiet" "-batch" "-nx" "-command=/checkout/obj/build/x86_64-unknown-linux-gnu/test/debuginfo/empty-string/empty-string.debugger.script"
[01:42:09] ------------------------------------------
[01:42:09] ------------------------------------------
[01:42:09] 0xb6fe1a30 in ?? ()
[01:42:09] Breakpoint 1 at 0xb6ff4560: file /checkout/src/test/debuginfo/empty-string.rs, line 31.
[01:42:09] 
[01:42:09] Breakpoint 1, empty_string::main::h2741c6f3b053ffe9 () at /checkout/src/test/debuginfo/empty-string.rs:31
[01:42:09] 31     zzz(); // #break
[01:42:09] $1 = {vec = {buf = {ptr = {pointer = 0x1 <error: Cannot access memory at address 0x1>, _marker = {<No data fields>}}, cap = 0, a = {<No data fields>}}, len = 0}}
[01:42:09] $2 = {data_ptr = 0xb6ff4d09 "", length = 0}
[01:42:09] A debugging session is active.
[01:42:09] 
[01:42:09]  Inferior 1 [process 22290] will be killed.
[01:42:09] 
[01:42:09] Quit anyway? (y or n) [answered Y; input not from terminal]
[01:42:09] ------------------------------------------
[01:42:09] stderr:
[01:42:09] ------------------------------------------
[01:42:09] ------------------------------------------
[01:42:09] warning: Missing auto-load script at offset 0 in section .debug_gdb_scripts
[01:42:09] of file /checkout/obj/build/x86_64-unknown-linux-gnu/test/debuginfo/empty-string/a.
[01:42:09] Use `info auto-load python-scripts [REGEXP]' to list them.
[01:42:09] warning: Unable to find dynamic linker breakpoint function.
[01:42:09] GDB will be unable to debug shared library initializers
[01:42:09] and track explicitly loaded dynamic code.
[01:42:09] warning: Unable to find dynamic linker breakpoint function.
[01:42:09] GDB will be unable to debug shared library initializers
[01:42:09] and track explicitly loaded dynamic code.
[01:42:09] warning: Could not load shared library symbols for 4 libraries, e.g. /system/bin/linker.
[01:42:09] Use the "info sharedlibrary" command to see the complete listing.
[01:42:09] Do you need "set solib-search-path" or "set sysroot"?
[01:42:09] ------------------------------------------
[01:42:09] 
[01:42:09] thread 'main' panicked at 'explicit panic', src/tools/compiletest/src/runtest.rs:3295:9
[01:42:09] note: Run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
---
[01:42:09] test result: FAILED. 71 passed; 1 failed; 48 ignored; 0 measured; 0 filtered out
[01:42:09] 
[01:42:09] 
[01:42:09] 
[01:42:09] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/arm-linux-androideabi/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--src-base" "/checkout/src/test/debuginfo" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/debuginfo" "--stage-id" "stage2-arm-linux-androideabi" "--mode" "debuginfo-both" "--target" "arm-linux-androideabi" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/checkout/obj/build/x86_64-unknown-linux-gnu/llvm/build/bin/FileCheck" "--linker" "/android/ndk/arm-14/bin/arm-linux-androideabi-clang" "--host-rustcflags" "-Crpath -O -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--target-rustcflags" "-Crpath -O -Zunstable-options  -Lnative=/checkout/obj/build/arm-linux-androideabi/native/rust-test-helpers" "--docck-python" "/usr/bin/python2.7" "--lldb-python" "/usr/bin/python2.7" "--llvm-version" "8.0.0\n" "--cc" "" "--cxx" "" "--cflags" "" "--llvm-components" "" "--llvm-cxxflags" "" "--remote-test-client" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/remote-test-client" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "/android/ndk/arm-14" "--color" "always"
[01:42:09] 
[01:42:09] 
[01:42:09] failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test --target arm-linux-androideabi
[01:42:09] Build completed unsuccessfully in 1:31:55
---
travis_time:end:011d8106:start=1551631444246113397,finish=1551631444253938006,duration=7824609
travis_fold:end:after_failure.3
travis_fold:start:after_failure.4
travis_time:start:0ba7c890
$ ln -s . checkout && for CORE in obj/cores/core.*; do EXE=$(echo $CORE | sed 's|obj/cores/core\.[0-9]*\.!checkout!\(.*\)|\1|;y|!|/|'); if [ -f "$EXE" ]; then printf travis_fold":start:crashlog\n\033[31;1m%s\033[0m\n" "$CORE"; gdb --batch -q -c "$CORE" "$EXE" -iex 'set auto-load off' -iex 'dir src/' -iex 'set sysroot .' -ex bt -ex q; echo travis_fold":"end:crashlog; fi; done || true
travis_fold:end:after_failure.4
travis_fold:start:after_failure.5
travis_time:start:18fece40
travis_time:start:18fece40
$ cat ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers || true
cat: ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers: No such file or directory
travis_fold:end:after_failure.5
travis_fold:start:after_failure.6
travis_time:start:010bc52a
$ dmesg | grep -i kill
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

euclio commented:
Amended the commit to ignore android as well.

alexcrichton commented:
@bors: r+

bors commented:
:pushpin: Commit 9e5def9616aba274a6c3138d9f66e778d50c35e0 has been approved by `alexcrichton`

<!-- @bors r=alexcrichton 9e5def9616aba274a6c3138d9f66e778d50c35e0 -->

bors commented:
:evergreen_tree: The tree is currently closed for pull requests below priority 50, this pull request will be tested once the tree is reopened
